File: backend/app/api/leaderboard.py
"""Leaderboard API endpoints.

## Overview

The leaderboard ranks contributors by $FNDRY earned. Features:
- **Time Periods**: Week, month, or all-time
- **Filters**: By tier, category
- **Top 3**: Extra metadata including medal, join date, best bounty

## Time Periods

| Period | Description |
|--------|-------------|
| week | Last 7 days |
| month | Last 30 days |
| all | All time (default) |

## Tier Filters

| Filter | Description |
|--------|-------------|
| 1 | Tier 1 bounties only |
| 2 | Tier 2 bounties only |
| 3 | Tier 3 bounties only |

## Category Filters

| Filter | Description |
|--------|-------------|
| frontend | Frontend work |
| backend | Backend work |
| security | Security work |
| docs | Documentation |
| devops | DevOps/Infrastructure |

## Response Fields

### Leaderboard Entry

| Field | Type | Description |
|-------|------|-------------|
| rank | integer | Position on leaderboard |
| username | string | GitHub username |
| display_name | string | Display name |
| avatar_url | string | Profile picture URL |
| total_earned | float | Total $FNDRY earned |
| bounties_completed | integer | Number of bounties |
| reputation_score | integer | Reputation points |
| wallet_address | string | Solana wallet address |

### Top 3 Metadata (for podium)

| Field | Type | Description |
|-------|------|-------------|
| medal | string | Medal emoji (🥇🥈🥉) |
| join_date | datetime | When they joined |
| best_bounty_title | string | Highest earning bounty |
| best_bounty_earned | float | Amount earned from best bounty |

## Caching

Results are cached for 60 seconds for performance.

## Rate Limit

100 requests per minute.
"""

from typing import Optional

from fastapi import APIRouter, Query

from app.models.leaderboard import (
    CategoryFilter,
    LeaderboardResponse,
    TierFilter,
    TimePeriod,
)
from app.services.leaderboard_service import get_leaderboard

router = APIRouter(prefix="/api", tags=["leaderboard"])


@router.get(
    "/leaderboard",
    response_model=LeaderboardResponse,
    summary="Get contributor leaderboard",
    description="""
Ranked list of contributors by $FNDRY earned.

## Features

- **Time Periods**: Filter by week, month, or all-time
- **Tier Filter**: Show only specific bounty tier earnings
- **Category Filter**: Show only specific category earnings
- **Top 3 Podium**: Extra metadata for top performers

## Example Requests

```
GET /api/leaderboard?period=week
GET /api/leaderboard?period=month&tier=1
GET /api/leaderboard?category=frontend&limit=50
```

## Response Structure

The response includes:
- `top3`: Top 3 contributors with extra metadata (medal, join date, best bounty)
- `entries`: All contributors including top 3 (for consistent display)
- `period`: Current time period filter
- `total`: Total number of contributors in result

## Caching

Results are cached for 60 seconds.

## Rate Limit

100 requests per minute.
""",
    responses={
        200: {
            "description": "Leaderboard data",
            "content": {
                "application/json": {
                    "example": {
                        "period": "all",
                        "total": 150,
                        "offset": 0,
                        "limit": 20,
                        "top3": [
                            {
                                "rank": 1,
                                "username": "topdev",
                                "display_name": "Top Developer",
                                "avatar_url": "https://avatars.githubusercontent.com/u/12345",
                                "total_earned": 50000.0,
                                "bounties_completed": 25,
                                "reputation_score": 1500,
                                "wallet_address": "ABC123...",
                                "meta": {
                                    "medal": "🥇",
                                    "join_date": "2024-01-01T00:00:00Z",
                                    "best_bounty_title": "Implement core escrow contract",
                                    "best_bounty_earned": 10000.0
                                }
                            }
                        ],
                        "entries": [
                            {
                                "rank": 1,
                                "username": "topdev",
                                "display_name": "Top Developer",
                                "avatar_url": "https://avatars.githubusercontent.com/u/12345",
                                "total_earned": 50000.0,
                                "bounties_completed": 25,
                                "reputation_score": 1500,
                                "wallet_address": "ABC123..."
                            }
                        ]
                    }
                }
            }
        }
    }
)
async def leaderboard(
    period: TimePeriod = Query(
        TimePeriod.all,
        description="Time period: week (last 7 days), month (last 30 days), or all (all-time)"
    ),
    tier: Optional[TierFilter] = Query(
        None,
        description="Filter by bounty tier: 1, 2, or 3"
    ),
    category: Optional[CategoryFilter] = Query(
        None,
        description="Filter by category: frontend, backend, security, docs, devops"
    ),
    limit: int = Query(20, ge=1, le=100, description="Results per page (max 100)"),
    offset: int = Query(0, ge=0, description="Pagination offset"),
) -> LeaderboardResponse:
    """Get ranked list of contributors by $FNDRY earned."""
    return get_leaderboard(
        period=period,
        tier=tier,
        category=category,
        limit=limit,
        offset=offset,
    )

File: frontend/src/api/leaderboard.ts
/**
 * Leaderboard API service.
 * Provides functions for fetching contributor leaderboard data.
 * @module api/leaderboard
 */

import { apiFetch, withRetry } from './client';
import type { Contributor, TimeRange } from '../types/leaderboard';

/** API response type for leaderboard entry */
interface ApiContributor {
  rank: number;
  username: string;
  avatar_url?: string;
  avatarUrl?: string;
  reputation_score?: number;
  points?: number;
  bounties_completed?: number;
  bountiesCompleted?: number;
  total_earned?: number;
  earningsFndry?: number;
  earnings_sol?: number;
  earningsSol?: number;
  streak?: number;
  top_skills?: string[];
  topSkills?: string[];
}

/** Map API contributor to frontend Contributor type */
function mapApiContributor(c: ApiContributor): Contributor {
  return {
    rank: c.rank,
    username: c.username,
    avatarUrl: c.avatar_url || c.avatarUrl || `https://api.dicebear.com/7.x/identicon/svg?seed=${c.username}`,
    points: c.reputation_score ? Math.floor(c.reputation_score * 100) : c.points || 0,
    bountiesCompleted: c.bounties_completed || c.bountiesCompleted || 0,
    earningsFndry: c.total_earned || c.earningsFndry || 0,
    earningsSol: c.earnings_sol || c.earningsSol || 0,
    streak: c.streak || Math.max(1, Math.floor((c.bounties_completed || 0) / 2)),
    topSkills: c.top_skills || c.topSkills || [],
  };
}

/** Time range mapping for API */
const TIME_RANGE_MAP: Record<TimeRange, string> = {
  '7d': '7d',
  '30d': '30d',
  '90d': '90d',
  all: 'all',
};

/** Fetch leaderboard contributors */
export async function fetchLeaderboard(
  timeRange: TimeRange = 'all',
  tier?: number,
  category?: string,
  limit: number = 50,
  offset: number = 0
): Promise<Contributor[]> {
  try {
    const params = new URLSearchParams();
    params.set('range', TIME_RANGE_MAP[timeRange]);
    params.set('limit', String(limit));
    params.set('offset', String(offset));
    
    if (tier) {
      params.set('tier', String(tier));
    }
    
    if (category) {
      params.set('category', category);
    }
    
    const response = await withRetry(() =>
      apiFetch<ApiContributor[]>(`/api/leaderboard?${params}`)
    );
    
    return response.map(mapApiContributor);
  } catch {
    return [];
  }
}

/** Fetch contributor by username or ID */
export async function fetchContributor(id: string): Promise<Contributor | null> {
  try {
    const response = await withRetry(() =>
      apiFetch<ApiContributor>(`/api/contributors/${id}`)
    );
    return mapApiContributor(response);
  } catch {
    return null;
  }
}